tests/integration/suite/scheduler/metrics/daprconnections.go

```diff
@@ -0,0 +1,121 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/dapr/dapr/tests/integration/framework"
+	"github.com/dapr/dapr/tests/integration/framework/process/daprd"
+	"github.com/dapr/dapr/tests/integration/framework/process/grpc/app"
+	"github.com/dapr/dapr/tests/integration/framework/process/scheduler"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(daprconnections))
+}
+
+type daprconnections struct {
+	scheduler *scheduler.Scheduler
+	daprdA    *daprd.Daprd
+	daprdB    *daprd.Daprd
+	daprdC    *daprd.Daprd
+}
+
+func (c *daprconnections) Setup(t *testing.T) []framework.Option {
+	c.scheduler = scheduler.New(t)
+	srv := app.New(t)
+
+	c.daprdA = daprd.New(t,
+		daprd.WithSchedulerAddresses(c.scheduler.Address()),
+		daprd.WithAppProtocol("grpc"),
+		daprd.WithAppPort(srv.Port(t)),
+	)
+
+	c.daprdB = daprd.New(t,
+		daprd.WithSchedulerAddresses(c.scheduler.Address()),
+		daprd.WithAppProtocol("grpc"),
+		daprd.WithAppPort(srv.Port(t)),
+	)
+
+	c.daprdC = daprd.New(t,
+		daprd.WithSchedulerAddresses(c.scheduler.Address()),
+		daprd.WithAppProtocol("grpc"),
+		daprd.WithAppPort(srv.Port(t)),
+	)
+
+	return []framework.Option{
+		framework.WithProcesses(srv, c.scheduler),
+	}
+}
+
+func (c *daprconnections) Run(t *testing.T, ctx context.Context) {
+	c.scheduler.WaitUntilRunning(t, ctx)
+
+	t.Run("ensure dapr connection with scheduler metric", func(t *testing.T) {
+		// 0 sidecars connected
+		metrics := c.scheduler.Metrics(t, ctx)
+		assert.Equal(t, 0, int(metrics["dapr_scheduler_sidecars_connected"]))
+
+		// 1 sidecar connected
+		c.daprdA.Run(t, ctx)
+		c.daprdA.WaitUntilRunning(t, ctx)
+		assert.EventuallyWithT(t, func(ct *assert.CollectT) {
+			metrics = c.scheduler.Metrics(t, ctx)
+			assert.Equal(ct, 1, int(metrics["dapr_scheduler_sidecars_connected"]))
+		}, 15*time.Second, 10*time.Millisecond, "daprdA sidecar didn't connect to Scheduler in time") //nolint:mnd
+
+		// 2 sidecars connected
+		c.daprdB.Run(t, ctx)
+		c.daprdB.WaitUntilRunning(t, ctx)
+		assert.EventuallyWithT(t, func(ct *assert.CollectT) {
+			metrics = c.scheduler.Metrics(t, ctx)
+			assert.Equal(ct, 2, int(metrics["dapr_scheduler_sidecars_connected"])) //nolint:mnd
+		}, 15*time.Second, 10*time.Millisecond, "daprdB sidecar didn't connect to Scheduler in time") //nolint:mnd
+
+		// 3 sidecars connected
+		c.daprdC.Run(t, ctx)
+		c.daprdC.WaitUntilRunning(t, ctx)
+		assert.EventuallyWithT(t, func(ct *assert.CollectT) {
+			metrics = c.scheduler.Metrics(t, ctx)
+			assert.Equal(ct, 3, int(metrics["dapr_scheduler_sidecars_connected"])) //nolint:mnd
+		}, 15*time.Second, 10*time.Millisecond, "daprdC sidecar didn't connect to Scheduler in time") //nolint:mnd
+
+		// 2 sidecars connected
+		c.daprdA.Cleanup(t)
```

Comment:
Please duplicate all daprd `Cleanup` function calls to just after the daprd `.Run` with a `t.Cleanup`. This ensures that daprds are _always_ cleandup, regardless of test failure. You'll need to add a `sync.Once` wrapper over the `exec.Cleanup` in the daprd framework process to ensure exec is attempted to kill twice.

```go
t.Cleanup(func() {c.daprdA.Cleanup(t)}())
```